### PR TITLE
Ignore Kamal files w/ Docker

### PR DESCRIPTION
**Pull request by ksylvest**
### Motivation / Background

The default configure includes both `.kamal/secrets` (may contain a `KAMAL_REGISTRY_PASSWORD` + `RAILS_MASTER_KEY`) and `config/deploy.yml`  in the built docker containers. This might make it easy to mistaken accidentally publish an image to a public container that contains credentials. This includes both the `.kamal` folder and `config/deploy.yml` file in  `.dockerignore` by default.

#### Without Change

```
docker build -t test .
docker run test cat .kamal/secrets
```

```
...
KAMAL_REGISTRY_PASSWORD=...
RAILS_MASTER_KEY=...
```

#### With Changes

```
docker build -t test .
docker run test cat .kamal/secrets
```

```
cat: .kamal/secrets: No such file or directory
```

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by jeromedalbert** [comment]
While `.kamal/secrets` should be safe for git in the [default setup](https://github.com/rails/rails/blob/a7c8a20825be5ac8d96566ec5dbe9daa85568bc8/railties/lib/rails/generators/rails/app/templates/kamal-secrets.tt#L3) and thus would be safe for Docker, if one veers off the default setup and puts raw credentials in there, they might remember to add `.kamal/secrets` to their `gitignore` but forget to add it in their `dockerignore`.

So I think this is a useful PR. And in any case, kamal-related files are not needed in a Docker container.

File: railties/lib/rails/generators/rails/app/templates/dockerignore.tt [hunk 1]
```diff
@@ -47,6 +47,12 @@
 /.github
 <% end -%>
 
+<% unless options.skip_kamal? -%>
+# Ignore Kamal files.
+/config/deploy.yml
+/.kamal
+<% end -%>
+
```
Comment (jeromedalbert) [inline railties/lib/rails/generators/rails/app/templates/dockerignore.tt:55]: Moving `<% end -%>` to the next line will will prevent a double empty line when Kamal is skipped:

```suggestion

<% end -%>
```

Comment (ksylvest) [reply railties/lib/rails/generators/rails/app/templates/dockerignore.tt]: Ah - fixed whitespace.

**Comment by zzak** [comment]
Could you add a test too please?

**Comment by ksylvest** [comment]
Sure - I've included a test for w/ `--skip-kamal` + w/o.

File: railties/lib/rails/generators/rails/app/templates/dockerignore.tt [hunk 2]
```diff
@@ -47,6 +47,12 @@
 /.github
 <% end -%>
 
+<% unless options.skip_kamal? -%>
+# Ignore Kamal files.
+/config/deploy.yml
```
Comment (jeromedalbert) [inline railties/lib/rails/generators/rails/app/templates/dockerignore.tt:52]: Should it be something like this so it also ignores kamal destinations like `config/deploy.staging.yml`?

```suggestion
/config/deploy*.yml
```

**Comment by ksylvest** [comment]
Hi @jeromedalbert, just wanted to check if any other changes are needed or what next steps are to get this over the finish line? Thanks!

**Comment by jeromedalbert** [comment]
Looks good on my end, let's see what the maintainers think.

In the meantime you could maybe rebase your branch see if the buildkite build goes green. The timezone failures seemed unrelated and it looks like they were just fixed on master a couple hours ago. You can also mark all my review comments as resolved.

**Comment by skipkayhil** [comment]
I think this PR duplicates (although more up to date and more complete) https://github.com/rails/rails/pull/47926. Could you add

```
Co-authored-by: Nicolai Reuschling <nicolai.reuschling@dkd.de>
```

to the bottom of the commit message so they are credited as well?

**Comment by ksylvest** [comment]
@skipkayhil I've included the co-author commit.

**Comment by ksylvest** [comment]
@skipkayhil @zzak  Just wanted to double check if any additional changes are required to get this reviewed / merged? Thanks!

**Pull request merged**